MSAL/src/native_auth/controllers/sign_up/MSALNativeAuthSignUpController.swift

```diff
@@ -625,51 +606,57 @@ final class MSALNativeAuthSignUpController: MSALNativeAuthBaseController, MSALNa
         }
     }
 
+    // swiftlint:disable:next function_body_length
     private func handleSubmitAttributesResult(
         _ result: MSALNativeAuthSignUpContinueValidatedResponse,
         username: String,
         signUpToken: String,
         event: MSIDTelemetryAPIEvent?,
         context: MSIDRequestContext
-    ) -> SignUpAttributesRequiredResult {
+    ) -> SignUpSubmitAttributesControllerResponse {
         switch result {
         case .success(let slt):
             let state = createSignInAfterSignUpStateUsingSLT(slt, username: username, event: event, context: context)
-            return .completed(state)
+            return .init(.completed(state), telemetryUpdate: { [weak self] result in
+                self?.stopTelemetryEvent(event, context: context, delegateDispatcherResult: result)
+            })
         case .attributesRequired(let signUpToken, let attributes):
             let error = AttributesRequiredError()
-            stopTelemetryEvent(event, context: context, error: error)
             MSALLogger.log(level: .error,
                            context: context,
                            format: "attributes_required received in signup/continue submitAttributes request: \(attributes)")
-
-            let state = SignUpAttributesRequiredState(controller: self,
-                                                      username: username,
-                                                      flowToken: signUpToken,
-                                                      correlationId: context.correlationId())
-
-            return .attributesRequired(attributes: attributes, state: state)
+            let state = SignUpAttributesRequiredState(
+                controller: self,
+                username: username,
+                flowToken: signUpToken,
+                correlationId: context.correlationId()
+            )
+            return .init(.attributesRequired(attributes: attributes, state: state), telemetryUpdate: { [weak self] result in
+                self?.stopTelemetryEvent(event, context: context, delegateDispatcherResult: result, controllerError: error)
```

Comment:
Hey Diego, can you help me understand this piece of code? We stop the telemetry using the delegateDispatcherResult that can be an error in case the external developer doesn't implements an optional delegate method. My question is about the controllerError. I suppose `error` is created at line 624, and it always be not nil, so this telemetry event will be always marked as a "failure" event, even if the external developer correctly implemented the optional delegate method. Is this behaviour expected?

Reply:
Yes, it is expected behavior. In this case, `.attributesRequired` is happening after the user has already sent some attributes before (the piece of code belongs to the function `handleSubmitAttributesResult()`.

 `.attributesRequired` can also happen after the user submits the code and submits the password. In those cases it is not considered an error, but a part of the flow. This was part of the original design.

So, now with optional delegates, this is what happens:

1. User sends some attributes, and receives that more attributes are needed (we consider this an error)
2. We check if the user has implemented `onSignUpAttributesRequired(attributes:newState)`
  a. If they have, we return in that method, and we finish the event with an error about attributes are needed.
  b. If they don't have this method, we return in `onSignUpAttributesRequiredError(error:)` saying that the user has to implement the other delegate. We also finish telemetry with the same error that says that the user needs to implement that method (in this case, the error about not having implemented the delegate is more important than not having sent some attributes)

Reply:
Ok, thanks for the clarification. Can you please add a comment that says that the telemetry event in this case will always be a failure because the user didn't submit all the required attributes. Thanks

Reply:
Added in those places where this happens.